rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,620 @@
+                parsed -= input_left;
+                input_left = 0;
+            }
+        }
+        return 1;
+    }
+}
+
+pub fn evaluate_stub_params(
+    input: &[u8],
+    input_len: u16,
+    hdrflags1: u8,
+    fraglenleft: u16,
+    stub_data_buffer: &mut Vec<u8>,
+    stub_data_buffer_len: &mut u16,
+) -> u16 {
+    let stub_len: u16;
+    stub_len = cmp::min(fraglenleft, input_len);
+    if stub_len == 0 {
+        return 0;
+    }
+    // If the frag is the the first frag irrespective of it being a part of
+    // a multi frag PDU or not, it indicates the previous PDU's stub would
+    // have been buffered and processed and we can use the buffer to hold
+    // frags from a fresh request/response
+    if hdrflags1 & PFC_FIRST_FRAG > 0 {
+        *stub_data_buffer_len = 0;
+    }
+
+    let input_slice = &input[..stub_len as usize];
+    stub_data_buffer.extend_from_slice(&input_slice);
+    *stub_data_buffer_len += stub_len;
+
+    stub_len
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_udp_parse(
+    _flow: *mut core::Flow,
+    state: &mut DCERPCUDPState,
+    _pstate: *mut std::os::raw::c_void,
+    input: *const u8,
+    input_len: u32,
+    _data: *mut std::os::raw::c_void,
+    _flags: u8,
+) -> i32 {
+    if input_len > 0 {
+        if input != std::ptr::null_mut() {
+            let buf = unsafe { std::slice::from_raw_parts(input, input_len as usize) };
+            return state.parse_dcerpc_udp(buf);
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_udp_state_free(state: *mut std::os::raw::c_void) {
+    let _drop: Box<DCERPCUDPState> = unsafe { transmute(state) };
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_udp_state_new() -> *mut std::os::raw::c_void {
+    let state = DCERPCUDPState::new();
+    let boxed = Box::new(state);
+    return unsafe { transmute(boxed) };
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_udp_state_transaction_free(
+    _state: *mut std::os::raw::c_void,
+    _tx_id: u64,
+) {
+    // do nothing
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_udp_get_tx_detect_state(
+    vtx: *mut std::os::raw::c_void,
+) -> *mut core::DetectEngineState {
+    let dce_state = cast_pointer!(vtx, DCERPCUDPState);
+    match dce_state.de_state {
+        Some(ds) => ds,
+        None => std::ptr::null_mut(),
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_udp_set_tx_detect_state(
+    vtx: *mut std::os::raw::c_void,
+    de_state: *mut core::DetectEngineState,
+) -> u8 {
+    let dce_state = cast_pointer!(vtx, DCERPCUDPState);
+    dce_state.de_state = Some(de_state);
+    0
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_udp_get_tx(
+    state: *mut std::os::raw::c_void,
+    _tx_id: u64,
+) -> *mut DCERPCUDPState {
+    let dce_state = cast_pointer!(state, DCERPCUDPState);
+    dce_state
+}
+
+#[no_mangle]
+pub extern "C" fn rs_dcerpc_udp_get_tx_cnt(_state: *mut std::os::raw::c_void) -> u8 {
+    1
```

Comment:
can we have multiple transactions in a single state? If so, this should return the highest tx id we've used (the API is misnamed, it doesn't return a count but instead a `max_id`)